docs/proposals/HealthChecks-L4.md

```diff
@@ -0,0 +1,287 @@
+    interval: 5s # how often health check requests should be made
+    unhealthy_threshold: 3
+    healthy_threshold: 2
+  passiveChecks:
+    unhealthy_threshold: 3
+    penalty_interval: 10s # for how long endpoint should be considered unhealthy
+```
+
+## Design
+
+### Requirement #1: Support `Envoy -> upstream` "health checks"
+
+* In order to generate `Cluster`s for a given `Dataplane`:
+  1. List all `HealthCheck`s in the same `mesh` as `Dataplane`
+  2. For each *outbound* interface of the `Dataplane`
+     * Go through `HealthCheck`s and select those where
+       * `destinations` selector matches `service` tag of that *outbound* interface
+       * `sources` selector matches tags on one of the *inbound* interfaces of that `Dataplane`
+     * Order matched `rule`s and keep only 1 "best match"
+     * If there is no "best match" `rule`
+       * Generate a `Cluster` without "active" and "passive" health checks
+     * If there is a "best match" `rule`
+       * Generate a `Cluster` with "active" and "passive" health checks according to that rule
+
+### Requirement #2: Support `Envoy -> local app` "health checks"
+
+TODO:
+
+### Requirement #3: Forward results of "health checks" to the Control Plane
+
+TODO:
+
+### Requirement #4: Aggregate results of "health checks" in the Control Plane
+
+TODO:
+
+### Requirement #5: Include "health statuses" into EDS configuration sent to Envoy
+
+TODO:
+
+## Consideration Notes
+
+### Requirement #3: Forward results of "health checks" to the Control Plane
+
+* Envoy already has a couple features we could build on top of, namely
+  * [event log](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/data/core/v3alpha/health_check_event.proto) for "active" health checks
+  * [event log](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/data/cluster/v2alpha/outlier_detection_event.proto) for "passive" health checks
+  * [Health xDS](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/discovery/v3alpha/hds.proto) for re-using Envoy as a worker node capable of doing health checks (completely unrelated to regular proxying)
+
+* event log for "active" health checks:
+  * at the moment, ONLY file output is supported
+    * see an example in the end
+  * configured individually for each `Cluster`
+
+* event log for "passive" health checks:
+  * at the moment, ONLY file output is supported
+    * see an example in the end
+  * configured globally for all `Cluster`s as part of `bootstrap` config
+
+* Health xDS:
+  * as part of this API, Envoy receives assignments and reports results back to the Control Plane over gRPC
+    * see an example in the end
+  * is not considered production ready yet (developed in summer 2018 by an intern)
+  * does not support mTLS
+
+Conclusions:
+* we can already use `Health xDS` for `Envoy -> local app` health checks
+* changes to the Envoy will be necessary to use `Health xDS` for `Envoy -> upstream` "health checks" (add support for mTLS)
```

Comment:
What is the use case of using HDS for `Envoy -> upstream`?  Wouldn't it be better to only HC your local app and send status to CP, which then updates list of endpoints for dataplanes that use this service?

Reply:
success of `Envoy -> local app` check doesn't give a full picture:
* it doesn't account for mTLS between client and server
* it doesn't account for different geographical location (e.g., connectivity to stand-by instances in another datacenter)